src/use-swr.ts

```diff
@@ -330,16 +305,74 @@ function useSWR<Data = any, Error = any>(
   const initialMountedRef = useRef(false)
 
   // do unmount check for callbacks
-  const eventsCallback = useCallback(
-    (event, ...params) => {
+  // do mounted check in suspense mode
+  // if key changed during the revalidation, old dispatch and config callback should not take effect.
+  const safeCallback = useCallback(
+    (callback: () => void) => {
       if (unmountedRef.current) return
       if (!initialMountedRef.current) return
       if (key !== keyRef.current) return
-      // @ts-ignore
-      configRef.current[event](...params)
+      callback()
     },
     [key]
   )
+  /**
+   * @param playload when you want to change stateRef, you should pass value explicitly
+   * @example
+   * ```js
+   * dispatch({
+   *   isValidating: false
+   *   data: newData // set data to newData
+   *   error: undefined // set error to undefined
+   * })
+   *
+   * dispatch({
+   *   isValidating: false
+   *   data: undefined // set data to undefined
+   *   error: Error // set error to Error
+   * })
+   * ```
+   */
+  const dispatch = useCallback(
+    (payload: Action<Data, Error>) =>
+      safeCallback(() => {
+        let shouldUpdateState = false
+        const { data, error, isValidating } = payload
+        const keys = Object.keys(payload)
+        if (
+          keys.includes('data') &&
+          !config.compare(stateRef.current.data, data)
+        ) {
+          stateRef.current.data = data
+          shouldUpdateState =
+            shouldUpdateState || stateDependencies.current.data
```

Comment:
if you update data to null/false/undefined, will `shouldUpdateState` still be false?

Reply:
That's `stateDependencies.current.data` (if the field is accessed by `render` or not) so should be OK.